lib/ansible/plugins/inventory/tower.py

```diff
@@ -0,0 +1,185 @@
+# Copyright (c) 2018 Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import (absolute_import, division, print_function)
+
+__metaclass__ = type
+
+DOCUMENTATION = '''
+    name: tower
+    plugin_type: inventory
+    authors:
+      - Matthew Jones (@matburt)
+      - Yunfan Zhang (@YunfanZhang42)
+    short_description: Ansible dynamic inventory plugin for Ansible Tower.
+    description:
+        - Reads inventories from Ansible Tower.
+        - Supports reading configuration from both YAML config file and environment variables.
+        - If reading from the YAML file, the file name must end with tower_inventory.(yml|yaml),
+          the path in the command would be /path/to/tower_inventory.(yml|yaml). If some arguments in the config file
+          are missing, this plugin will try to fill in missing arguments by reading from environment variables.
+        - If reading configurations from environment variables, the path in the command must be @tower_inventory.
+    options:
+        plugin:
+            description: the name of this plugin, it should always be set to 'tower'
+                for this plugin to recognize it as it's own.
+            env:
+                - name: ANSIBLE_INVENTORY_ENABLED
+            required: True
+            choices: ['tower']
+        tower_host:
+            description: The network address of your Ansible Tower host.
+            type: string
+            env:
+                - name: TOWER_HOST
+            required: True
+        tower_username:
+            description: The user that you plan to use to access inventories on tower.
+            type: string
+            env:
+                - name: TOWER_USERNAME
+            required: True
+        tower_password:
+            description: The password for tower_username.
+            type: string
+            env:
+                - name: TOWER_PASSWORD
+            required: True
+        tower_inventory:
+            description: The ID of the tower inventory you wish to import.
+            type: string
+            env:
+                - name: TOWER_INVENTORY
+            required: True
+        tower_verify_ssl:
+            description: Specify whether Ansible should verify the SSL certificate of Ansible tower.
+            type: bool
+            default: True
+            env:
+                - name: TOWER_VERIFY_SSL
+            required: False
+'''
+
+EXAMPLES = '''
+# Before you execute the following commands, you should make sure this file is in your plugin path,
+# and you enabled this plugin.
+
+# Example for using tower_inventory.yml file
+
+plugin: tower
+tower_host: your_tower_server_network_address
+tower_username: your_username
+tower_password: your_password
```

Comment:
I'm sorry if I'm being a broken record about this, but I see lots of hazards for us in the future. Maintaining this could be a large undertaking if there's not a good deal of congruency. I now see _three_ different forms of configuration files:

### inventory plugin

```yaml
plugin: tower
tower_host: https://localhost:8043/
tower_username: admin
tower_password: password1
tower_inventory: 6
tower_verify_ssl: False
```

### tower modules

(docs actually call out `key=value` syntax, but docs are poorly thought out in several ways currently)

parameters match what are directly passed into the modules

```yaml
host: https://localhost:8043/
username: admin
password: password1
verify_ssl: False
```

### tower-cli config

```configparser
[general]
host = http://localhost:8013/
verify_ssl = false
oauth_token = xxxxxxxxxxxxxxx
```

For the most part, I agree with the environment variable keys used here. However, I know this is painful, but it seems that the inventory config format is the odd-one out here. The tower-cli code can be adjusted to accept the YAML format. The modules can be free to ignore other variables like "plugin", but we can't go back once we've introduced the long names in the inventory plugin config.